### ISSUE TITLE
Assembly conflicts during compilation

### ISSUE THREAD

**Issue by daveaglick** [cake-build/cake#2134]
### What You Are Seeing?

@redth reports the following messages during script compilation:
```
error CS1705: Assembly 'Cake.Json' with identity 'Cake.Json, Version=3.0.0.0, Culture=neutral, PublicKeyToken=null' uses 'Newtonsoft.Json, Version=11.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' which has a higher version than referenced assembly 'Newtonsoft.Json' with identity 'Newtonsoft.Json, Version=9.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed'
```

And then the compilation fails with:

```
Error: Cake.Core.CakeException: Error(s) occurred when compiling build script:
  at Cake.Scripting.Roslyn.RoslynScriptSession.Execute (Cake.Core.Scripting.Script script) [0x00182] in <f2dfd1d7ff874bb5925674ed0d0fc83d>:0 
  at Cake.Core.Scripting.ScriptRunner.Run (Cake.Core.Scripting.IScriptHost host, Cake.Core.IO.FilePath scriptPath, System.Collections.Generic.IDictionary`2[TKey,TValue] arguments) [0x00358] in <bf8648e2e0a54f859ebcbb1728e8c9cf>:0 
  at Cake.Commands.BuildCommand.Execute (Cake.CakeOptions options) [0x00026] in <f2dfd1d7ff874bb5925674ed0d0fc83d>:0 
  at Cake.CakeApplication.Run (Cake.CakeOptions options) [0x00015] in <f2dfd1d7ff874bb5925674ed0d0fc83d>:0 
  at Cake.Program.Main () [0x000d1] in <f2dfd1d7ff874bb5925674ed0d0fc83d>:0 
```

### What is Expected?

Successful compilation.

### What version of Cake are you using?

0.27.1

### Output Log

https://gist.github.com/Redth/37bfa030e2890d22d93c06b154648e72

---

Have been discussing in the Gitter. This is related to #2116 (but not quite the same). We [added](https://github.com/cake-build/cake/blob/52bd7efc142f47887bdd8d73ef8f508e12c5d654/src/Cake.Core/Cake.Core.csproj#L19) a `PackageReference` to `Cake.Core` to explicitly reference the latest `Newtonsoft.Json` 11.0.2, but only for .NET Standard since that's the profile where it was being transitively pulled in from via `Microsoft.Extensions.DependencyModel`. _However_, `Newtonsoft.Json` is also pulled if transitively from `Cake.NuGet` which is still referencing an older version of `Newtonsoft.Json`. That causes the .NET Framework builds of Cake to continue placing the older version of `Newtonsoft.Json` in the output because the .NET Standard `PackageReference` from `Cake.Core` isn't applicable.

The solution will be to add additional `PackageReference` to the latest `Newtonsoft.Json`. We could do a couple things:
* Remove the .NET Standard reference from `Cake.Core` and add a reference to both profiles in `Cake`. If we do this, we'll need to double-check it still takes effect for `Cake.CoreCLR` and the newer version of  `Newtonsoft.Json.dll` ends up in the output for both builds.
* Add an additional `PackageReference` for `Newtonsoft.Json` to `Cake.NuGet` that applies to both profiles.

This also exposes a potentially more serious issue: we can guard against many/most instances of this problem by continuing to bump the JSON.NET reference with every Cake release, but in the intermediate time if an addin references a version higher than what we ship...boom. The answer to this broader problem will be more complicated. One approach might be accumulating all the assembly references because loading them in the `ScriptRunner` and then loading the highest version.

**Comment by mholo65** [comment]
@daveaglick I'd like to see an approach of reflection only loading the addins during pre-processing stage, and then create the dependency graph using highest-win-model.

Problem will however always be the dependencies we ship with Cake... Autofac, Roslyn, NuGet (and their dependencies)....

**Comment by mholo65** [comment]
...or even executing the cake script in a separate App domain / Assembly load context (if that's possible)...

**Comment by daveaglick** [comment]
@mholo65 I agree reflection-only loading to build the dependency graph it the best long-term solution. Whenever I've played with AppDomains in the past I've always wished I had those hours back. The new custom context stuff in .NET Core looks interesting, but it's .NET Core only and we probably don't want to go down the route of drastically different logic depending on platform.
